pareto/operational_water_management/operational_produced_water_optimization_model.py

```diff
@@ -3831,3 +3879,769 @@ def postprocess_water_quality_calculation(model, df_sets, df_parameters, opt):
+            model.s_S,
+            model.s_T,
+            model.s_W,
+            model.s_Q,
+            rule=lambda model, s, t, w, q: model.v_F_DiscreteFlowOutStorage[s, t, w, q]
+            <= (
+                model.p_sigma_Storage[s]
+                + sum(
+                    model.p_sigma_Pipeline[s, n] + max(model.p_delta_Pipeline.values())
+                    for n in model.s_N
+                    if model.p_SNA[s, n]
+                )
+                + sum(
+                    model.p_sigma_Pipeline[s, p] + max(model.p_delta_Pipeline.values())
+                    for p in model.s_CP
+                    if model.p_SCA[s, p]
+                )
+                + sum(
+                    model.p_sigma_Pipeline[s, k] + max(model.p_delta_Pipeline.values())
+                    for k in model.s_K
+                    if model.p_SKA[s, k]
+                )
+                + sum(
+                    model.p_sigma_Pipeline[s, r] + max(model.p_delta_Pipeline.values())
+                    for r in model.s_R
+                    if model.p_SRA[s, r]
+                )
+                + sum(
+                    model.p_sigma_Pipeline[s, o] + max(model.p_delta_Pipeline.values())
+                    for o in model.s_O
+                    if model.p_SOA[s, o]
+                )
+                + sum(
+                    (model.p_delta_Truck * model.p_max_number_of_trucks)
+                    for p in model.s_CP
+                    if model.p_SCT[s, p]
+                )
+                + sum(
+                    (model.p_delta_Truck * model.p_max_number_of_trucks)
+                    for k in model.s_K
+                    if model.p_SKT[s, k]
+                )
+            )
+            * model.v_Z[s, t, w, q],
+            doc="Only one outflow for storage site s can be non-zero for quality component w and all discretized quality q",
+        )
+
+        model.SumDiscreteFlowsIsFlowOutStorage = Constraint(
+            model.s_S,
+            model.s_T,
+            model.s_W,
+            rule=lambda model, s, t, w: sum(
+                model.v_F_DiscreteFlowOutStorage[s, t, w, q] for q in model.s_Q
+            )
+            == (
+                model.v_L_Storage[s, t]
+                + sum(model.v_F_Piped[s, n, t] for n in model.s_N if model.p_SNA[s, n])
+                + sum(model.v_F_Piped[s, p, t] for p in model.s_CP if model.p_SCA[s, p])
+                + sum(model.v_F_Piped[s, k, t] for k in model.s_K if model.p_SKA[s, k])
+                + sum(model.v_F_Piped[s, r, t] for r in model.s_R if model.p_SRA[s, r])
+                + sum(model.v_F_Piped[s, o, t] for o in model.s_O if model.p_SOA[s, o])
+                + sum(
+                    model.v_F_Trucked[s, p, t] for p in model.s_CP if model.p_SCT[s, p]
+                )
+                + sum(
+                    model.v_F_Trucked[s, k, t] for k in model.s_K if model.p_SKT[s, k]
+                )
```

Comment:
Does this constraint  also need to consider the water in storage in the previous time period ( model.v_L_Storage[s, model.s_T.prev(t)] or odel.p_lambda_Storage[s], initial storage)